### PR TITLE
Issue #6208: add hook_field_schema_alter()

### LINKED ISSUE #6208: Add hook_field_schema_alter()
## Description of the need

Add `hook_field_schema_alter()`, as described in https://www.drupal.org/node/691932, which has been merged into Drupal 7 core as of 7.99. See https://www.drupal.org/node/3406312

For the reasons we have hook_schema_alter() - mainly site-specific tweaks, we should allow the same for field storage.

It would make it easier to implement a contrib module (like `date_repeat`) that would alter core fields (like core date fields).

## Proposed solution

Apply this patch: https://www.drupal.org/files/issues/2023-03-16/drupal-field_schema_alter-691932-129-D7.patch or similar.

## Alternatives that have been considered

`hook_schema_alter()` can be used too but it requires checking for each table to see if it's of the right field type and also requires also handling field creation and deletion so it's more work than the proposed new hook.

## Example implementation

```
/**
 * Allow modules to alter the schema for a field.
 *
 * @param array $schema
 *   The schema definition as returned by hook_field_schema().
 * @param array $field
 *   The field definition.
 *
 * @see field_retrieve_schema()
 */
function hook_field_schema_alter(&$schema, $field) {
  if ($field['type'] == 'image') {
    // Alter the length of a field.
    $schema['columns']['alt']['length'] = 2048;
    // Add an additional column of data.
    $schema['columns']['additional_column'] = array(
      'description' => "Additional column added to image field table.",
      'type' => 'varchar',
      'length' => 128,
      'not null' => FALSE,
    );
    // Add an additional index.
    $schema['indexes']['fid_additional_column'] = array('fid', 'additional_column');
  }
}
````

### PR DESCRIPTION
**Pull request by herbdool**
Fixes https://github.com/backdrop/backdrop-issues/issues/6208

### CODE REVIEW

File: core/modules/field/tests/field_test_schema_alter/field_test_schema_alter.install [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?php
+/**
+ * @file
+ * Install, update and uninstall functions for the field_test_schema_alter module.
```
Comment (bugfolder) [inline core/modules/field/tests/field_test_schema_alter/field_test_schema_alter.install]:
```suggestion
 * Install, update and uninstall functions for the field_test_schema_alter
 * module.
```

To fix line length.

**Comment by herbdool** [comment]
Unclear why the layout tests are failing and only in PHP 5.6 and 8.2. Perhaps random?

**Comment by bugfolder** [comment]
This is a known problem that afflicts more and and more PRs these days. You can ignore testLayoutUpgrade fails (unless you're actually doing a layout-related PR.) There's an issue for it: https://github.com/backdrop/backdrop-issues/issues/6293

**Comment by bugfolder** [comment]
Theres an issue: https://github.com/backdrop/backdrop-issues/issues/6293

**Comment by herbdool** [comment]
@bugfolder feel like reviewing this? Now that it's in Drupal 7.99 release makes it easier to port here.

**Pull request merged**